### ISSUE TITLE
tf.reduce_*(mean/sum) runs very slow on GPU

### ISSUE THREAD

**Issue by vs-zhehangd** [tensorflow/tensorflow#13171]
I notice that tf.reduce_*(mean/sum) runs very slow on GPU in some cases, which can happen in the following simple examle:

x = tf.Variable(tf.ones([80, 80, 80, 80])) # 4-D tensor.
y = tf.reduce_sum(x, [0, 2, 3]) # Sum over all dims except the 2nd.

The execution time on GPU is very large and is approximate same as (or more than) the time on CPU, which probably means the GPU is not used at all. The same result can be obtained by choosing the other axes, except for the first and the last axes, in which case the execution on GPU is significantly faster than CPU.

Here is the [code](https://gist.github.com/vs-zhehangd/8a547094cfa0efc181b814bfb20b31ce) that reproduces the problem. You can run with `--keep_dim k` k=0,1,2,3 to select different axes.

I am using the following PC system:

* Kubuntu 16.04
* TensorFlow 1.2.1 ('v1.2.0-5-g435cdfc', '1.2.1')
* Python 2.7
* GeForce GTX 1080 Ti
* CUDA-8.0

The execution time on GPU, CPU, and NumPy is given as follows:

 exec time (s) | [1,2,3] | [0,2,3] | [0,1,3] | [0,1,2] |
|---------------|---------|---------|---------|---------|
| GPU           | 0.00918 | 0.40572 | 0.55388 | 0.01905 |
| CPU           | 0.05921 | 0.22461 | 0.56524 | 0.16172 |
| NumPy         | 0.24799 | 0.24847 | 0.24886 | 0.26601 |

Similar result was observed also on my laptop.

-------------------------------------------------------

I met this problem when I tried to implement Batch Normalization (initially when I ran tf.nn.moment(), then realized the key was tf.reduce_*). I wanted to collect means of different channels. Using NHWC data format the problem does not matter as C is the last dimension. However, slow execution occurs if I want to implement for NCHW format as C is the second dimension, this makes the execution time on Batch Normalization overwhelms all other ops such as convolutions. This is surely a nightmare for training and evaluation.

I hope it will be fixed if it is a bug. Or, if the reason is that the oprations are just not implemented for GPU by now, I wonder if there is a way to walk around.

**Comment by yaroslavvb** [comment]
I have GTX 1080 and get much better numbers using TF head from Sep 13

(sep13) yaroslav@home:~/git0/stuff$ python reduce_sum_runs_slow.py --keep_dim 0
np time    0.20407
tf time    0.01083

(sep13) yaroslav@home:~/git0/stuff$ python reduce_sum_runs_slow.py --keep_dim 1
np time    0.21019
tf time    0.01631

(sep13) yaroslav@home:~/git0/stuff$ python reduce_sum_runs_slow.py --keep_dim 2
np time    0.20842
tf time    0.02032

(sep13) yaroslav@home:~/git0/stuff$ python reduce_sum_runs_slow.py --keep_dim 3
np time    0.33465
tf time    0.01350

**Comment by vs-zhehangd** [comment]
I tested the benchmark program on four different machines (one uses 0.12 and the rest all use 1.2r) and the problem all occurs. I upgraded TensorFlow from 1.2 to 1.3 via pip on my main server and the problem remains the same.

Then I compiled the latest source cloned this morning. it seems that this problem has been fixed. : )

**Comment by yaroslavvb** [comment]
It's easy to test with latest version, no recompilation needed. IE, get the pip link from tensorflow github  front page, then

```
conda create --name sep13 --clone my_previous_env
source activate sep13
pip install --upgrade https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3.5,label=gpu-linux/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow_gpu-1.3.0-cp35-cp35m-linux_x86_64.whl
python run_my_script.py
```

**Comment by frank2wang87** [comment]
@yaroslavvb  I got the same problem on tf 1.3.0-rc2(installed by pip install). Does it mean that I should compile the source code?

**Comment by inders** [comment]
I am also hitting this problem. What is the suggested solution or workarounds if any.

**Comment by yaroslavvb** [comment]
It's been fixed, so upgrade to latest version
